### PR TITLE
New command for verifying VPC's DNS config

### PR DESCRIPTION
**Pull request by abyrne55**
This PR adds a new command, `dns`, that verifies that a given VPC meets the criteria spelled out in the [OpenShift Docs](https://docs.openshift.com/container-platform/4.10/installing/installing_aws/installing-aws-vpc.html). In practice, this means ensuring that the attributes `enableDnsSupport` and `enableDnsHostnames` for the given VPC are both true.

This addresses [OSD-9565](https://issues.redhat.com/browse/OSD-9565), which says that having a way to quickly rule out DNS misconfigurations would greatly speed up VPC debugging efforts.

### CODE REVIEW

File: cmd/dns/cmd.go [hunk 1]
```diff
@@ -50,7 +50,8 @@ func NewCmdValidateDns() *cobra.Command {
 
 			logger.Warn(ctx, "Using region: %s", config.region)
 			creds := credentials.NewStaticCredentialsProvider(os.Getenv("AWS_ACCESS_KEY_ID"), os.Getenv("AWS_SECRET_ACCESS_KEY"), os.Getenv("AWS_SESSION_TOKEN"))
-			cli, err := cloudclient.NewClient(ctx, logger, creds, config.region, "", nil)
+			// The use of t3.micro here is arbitrary; we just need to provide any valid machine type
+			cli, err := cloudclient.NewClient(ctx, logger, creds, config.region, "t3.micro", nil)
```
Comment (fahlmant) [inline cmd/dns/cmd.go:54]: This is out of scope of your PR, but we should really remove this requirement... its a bit silly.

File: pkg/cloudclient/aws/private.go [hunk 2]
```diff
@@ -424,7 +424,40 @@ func (c *Client) validateEgress(ctx context.Context, vpcSubnetID, cloudImageID s
...
+	dnsSprtInput := &ec2.DescribeVpcAttributeInput{
+		Attribute: "enableDnsSupport",
+		VpcId:     aws.String(vpcID),
+	}
+	dnsHostInput := &ec2.DescribeVpcAttributeInput{
+		Attribute: "enableDnsHostnames",
+		VpcId:     aws.String(vpcID),
+	}
+
+	dnsSprtResult, dnsSprtErr := c.ec2Client.DescribeVpcAttribute(ctx, dnsSprtInput)
```
Comment (fahlmant) [inline pkg/cloudclient/aws/private.go:438]:
```suggestion
	dnsSprtResult, dnsSprtErr := c.ec2Client.DescribeVpcAttribute(ctx,  &ec2.DescribeVpcAttributeInput{
		Attribute: "enableDnsSupport",
		VpcId:     aws.String(vpcID),
	})
```

File: pkg/cloudclient/aws/private.go [hunk 3]
```diff
@@ -424,7 +424,40 @@ func (c *Client) validateEgress(ctx context.Context, vpcSubnetID, cloudImageID s
...
+	dnsHostResult, dnsHostErr := c.ec2Client.DescribeVpcAttribute(ctx, dnsHostInput)
```
Comment (fahlmant) [inline pkg/cloudclient/aws/private.go:439]:
```suggestion
	dnsHostResult, dnsHostErr := c.ec2Client.DescribeVpcAttribute(ctx, &ec2.DescribeVpcAttributeInput{
		Attribute: "enableDnsHostnames",
		VpcId:     aws.String(vpcID),
	})
```

File: pkg/cloudclient/aws/private.go [hunk 4]
```diff
@@ -424,7 +424,40 @@ func (c *Client) validateEgress(ctx context.Context, vpcSubnetID, cloudImageID s
...
+		c.logger.Error(ctx, "DNS support is disabled for VPC %s", vpcID)
+	}
+	if !*dnsHostResult.EnableDnsHostnames.Value {
+		verificationFailed = true
+		c.logger.Error(ctx, "DNS hostnames are disabled for VPC %s", vpcID)
+	}
+	// Report final result
+	if verificationFailed {
+		c.logger.Error(ctx, "VPC %s's DNS configuration is incompatible with OSD. Please correct the errors above and try again.", vpcID)
+	}
```
[the remark this answers is not in the document]
Comment (fahlmant) [reply pkg/cloudclient/aws/private.go:460]: Thoughts on this?

File: cmd/dns/cmd.go (hunk 1 above)
Comment (abyrne55) [reply cmd/dns/cmd.go:54]: Completely agree. Tried to leave this field blank, but it refused to instantiate unless a valid type was provided

Comment (fahlmant) [reply cmd/dns/cmd.go:54]: I can work on a separate PR

File: pkg/cloudclient/aws/private.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (abyrne55) [reply pkg/cloudclient/aws/private.go:460]: Agreed, I much prefer the succinct version. I'll be committing this fix (plus fixes for the failed tests) shortly

**Comment by abyrne55** [comment]
This PR is now ready for review, but I'd like to note 2 nits that I'm aware of but feel should be addressed in separate PRs.

- As @fahlmant noted, `cloudclient.NewClient` should not require an instance type to be specified, as not all commands using this interface are creating instances
- `Client.output.SetFailures()` records input failures using `NewEgressURLError()`, which is awkward for every command that's not `egress`, because this happens:
```sh
/osd-network-verifier dns --vpc-id=vpc-XYZ
Using region: us-east-1
Verifying DNS config for VPC vpc-XYZ
DNS Support for VPC vpc-XYZ: true
DNS Hostnames for VPC vpc-XYZ: false
Both DNS support and DNS hostnames must be enabled on VPC vpc-XYZ in order to be compatible with OSD.
Summary:
printing out failures:
 -  egressURL error: VPC DNS verification failed
printing out exceptions preventing onv from running:
printing out errors faced during the execution:
Failure!
```

**Comment by fahlmant** [comment]
/hold
This looks great as is. I'm placing a hold as I believe there's another requirement around the DHCP configuration of the VPC, so I'd like to check that first.

**Comment by bdematte** [comment]
/retest
Fixed the formating issues

**Comment by fahlmant** [comment]
/hold cancel
We can add the other part of this test later if needed

**Review by fahlmant — APPROVED** [review]
/approve
/lgtm

**Pull request merged**